### PR TITLE
Cypress POC

### PR DESCRIPTION
**Pull request by cconard96**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -

Suggested alternative to Symfony/Panther (#14999).
This PR currently contains no meaningful/real tests. It **does** however show that it is capable of some things that were difficult/impossible with Panther like typing within a TinyMCE editor or interacting with flatpickr.
Also, to say docker/github actions are something I'm not good with would be an understatement. So, the base URL is hard-coded for my development instance currently in the config rather than starting GLPI in a container using a test database. The tests done currently don't make any changes to the data except changing your current ticket search. **I will need help getting it working in a containerized environment like the other tests.**

- Flatpickr was handled with a custom `selectDate` command. This command can select the date by either interacting with flatpickr itself (default) or simply typing the date depending on the "interactive" option.
- TinyMCE was handled by hooking into Cypress' `type` command and detecting if it is a textarea with a tinymce editor. I added examples of using the `setContent` function on the editor as well as interactively typing within the editor.
- Select2 works completely without issue as long as you use the "force" option with the `select` command. The "force" option tells Cypress to ignore the visibility checks it does. Another option could be hooking into the command like I did for `type` to detect select2 and handle it without requiring the "force" option.

Finally, I added a `test:e2e` npm command which runs Cypress in interactive mode (`cypress open`). The tests can be run in headless mode with `cypress run` instead. When the Cypress window opens, you will need to select the "E2E" option. When the tests are run in interactive/headed mode, you can navigate through each test/step and see a screenshot of what the screen looked like at the time.

`cypress install`
`npm run test:e2e`
- Select E2E Testing
- Pick the browser you want to run the tests in. Depends on what you have installed. My options were Chrome, Electron and Firefox. There is a way to add experimental Webkit (Safari) support later if desired through `playwright-webkit`.
- Select the "poc.cy.js" spec
- Wait and watch the tests run. You can see the actual browser window as Cypress runs the tests and interacts with it.
- In the second panel on the left you can see a list of the tests that were run. You can click them to expand them to see the individual steps. Hovering over the steps will show you a preview of what the screen looked like (not perfect).

### CODE REVIEW

**Comment by trasher** [comment]
I took a look; this seems promising.

It does not work on my side, but I guess this is a miss of data (we should ensure we're using an instance with existing bootstrapped data from test suite).

I won't be able to help with actions automation.

**Comment by cconard96** [comment]
> I took a look; this seems promising.
> 
> It does not work on my side, but I guess this is a miss of data (we should ensure we're using an instance with existing bootstrapped data from test suite).

The only thing that should need changed to test manually is the baseURL config option.

For automated testing, I think we would need a container that actually runs a web server. Currently, we just use the `-fpm` PHP docker image. If we replace that with `-apache` for the app container, we should be able to use the Cypress GitHub action to run the tests.
https://github.com/cypress-io/github-action

**Comment by trasher** [comment]
> > I took a look; this seems promising.
> > It does not work on my side, but I guess this is a miss of data (we should ensure we're using an instance with existing bootstrapped data from test suite).
> 
> The only thing that should need changed to test manually is the baseURL config option.

Using a fresh install launched with a `php -S`, cypress suite is currently failing on my side: `GLPIDashboard is not defined`.

**Comment by cconard96** [comment]
POC ready for review. Obviously there were some changes made to the GH actions to focus just on the E2E tests.
At the very least, you can review the code and see that it actually works in the GH actions.

I don't know why, but during my work on this PR the CIs were suddenly failing to get the `plain-scrollbar` library. As a workaround, I dropped the library and implemented a "fake" scrollbar for the entity selector myself. Not a perfect implementation, but works for the one place that uses it.

Would need adjusted:
- Since we are actually using GLPI in the tests, we have to use the real configs rather than "tests/configs". This makes testing locally complicated since you wouldn't want to mess with your development/production instance.
- Using a PHP Apache container (which would run Cypress as well) instead of FPM. Currently, this is published in my organization so it would need changed to be in the official GLPI docker images repo.

**Comment by cedric-anne** [comment]
> I don't know why, but during my work on this PR the CIs were suddenly failing to get the `plain-scrollbar` library. As a workaround, I dropped the library and implemented a "fake" scrollbar for the entity selector myself. Not a perfect implementation, but works for the one place that uses it.

Maybe the issue was related to some rate limiting on Github side.
Anyway, as the `ewya/PlainScrollbar` package does not seems to be maintained, if the performances of your own implementation is OK with a huge entity tree, then I guess we could use it as a replacement. Please open a separate PR for this.

> * Since we are actually using GLPI in the tests, we have to use the real configs rather than "tests/configs". This makes testing locally complicated since you wouldn't want to mess with your development/production instance.

We could probably adapt the default `GLPI_CONFIG_DIR` value to `tests/config` when `GLPI_ENVIRONMENT_TYPE=testing`. This value is already defined in the container image used by CI, see
https://github.com/glpi-project/docker-images/blob/146c8eaec1aedd4aa7007ecfbcbcf7b91177bbf5/githubactions-php/Dockerfile#L130-L132
I think it would be the best solution for people that would run the tests using the `tests/run_tests.sh` script.
For the people that would like to run the tests locally without using docker, they could probably use a distinct local port that points to a apache/nginx/whaterver host that defines the `GLPI_ENVIRONMENT_TYPE=testing` constant.

> * Using a PHP Apache container (which would run Cypress as well) instead of FPM. Currently, this is published in my organization so it would need changed to be in the official GLPI docker images repo.

Could you open a PR for this on the `glpi-project/docker-images` repository?

**Comment by cedric-anne** [comment]
@cconard96 

With the following patch, the Apache server is no longer a requirement and there is no more issue with the config file:
```diff
diff --git a/.github/actions/test_tests-e2e.sh b/.github/actions/test_tests-e2e.sh
old mode 100644
new mode 100755
index 08feb36773..dfa651ef74
--- a/.github/actions/test_tests-e2e.sh
+++ b/.github/actions/test_tests-e2e.sh
@@ -1,14 +1,12 @@
 #!/bin/bash
 set -e -u -x -o pipefail
 
-# Copy the configs created by the install command "tests/config" to "config"
-cp -r tests/config/. config
-
 # Set baseUrl for Cypress in environment variable
-export CYPRESS_BASE_URL="http://localhost:80"
+export CYPRESS_BASE_URL="http://localhost:8088"
 
 # Install Cypress
 node_modules/.bin/cypress install
 
 # Run Cypress tests
+php -S localhost:8088 -t public tests/router.php &>/dev/null &
 node_modules/.bin/cypress run --project tests
```

**Comment by cedric-anne** [comment]
I pushed a commit to use the new CI image built with Apache and Cypress (see https://github.com/glpi-project/docker-images/pull/118). I adapted the config defauts to the environment type, to simplify configuration files handling (and to unlock the ability to execute e2e tests from `run_tests.sh` script. It seems to work as expected.

**Comment by cconard96** [comment]
I've re-enabled the other CI jobs and re-enabled the POC test done with the ticket search. If tests still pass, I think this is ready for final validation. Actual tests can be added in later PRs.

File: inc/based_config.php [hunk 1]
```diff
@@ -85,16 +85,21 @@ function (string $included_file) {
...
+
+    // Define GLPI_* constants that can be customized by admin.
+    //
+    // Use a self-invoking anonymous function to:
+    // - prevent any global variables/functions definition from `local_define.php` and `downstream.php` files;
+    // - prevent any global variables definition from current function logic.
 
     $constants = [
+        // Env type
+        'GLPI_ENVIRONMENT_TYPE' => 'production',
```
[the remark this answers is not in the document]
Comment (cedric-anne) [reply inc/based_config.php:98]: This is the value. It will only b used if this constant is not set in the current environment.

**Comment by cconard96** [comment]
> One of the tests doesn't seem to pass on my side:

As long as it is using the test database and bootstrap file to load some initial data, it shouldn't be showing no tickets here. It isn't required to bootstrap initial data though; I think creating the items manually would be better for test coverage.

I don't think it is an issue to remove the tests currently in this PR anyways since they just show a proof of concept and show how testing certain elements that we had difficulty testing with Panther can be tested with Cypress.

File: tests/cypress/support/commands.js [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    );
+});
+
+/**
+ * Wait for libraries to replace inputs with their own components
+ */
+Cypress.Commands.add('waitForInputs', () => {
+    // eslint-disable-next-line cypress/no-unnecessary-waiting
+    cy.wait(1000);
+});
```
[the remark this answers is not in the document]
Comment (cconard96) [reply tests/cypress/support/commands.js:94]: It's just a matter of using cy.get which will already wait a certain amount of time for the element to exist, for most things. The main issue though are the elements like drop-down, rich text boxes and datetime which get initialized client-side after the element already exists. We would need to identify some extra DOM elements that are added after initialization and use cy.get on them.

[the remark this answers is not in the document]
Comment (AdrienClairembault) [reply tests/cypress/support/commands.js:94]: One test fail on my machine too for some reason (but work on the CI).
With this changes, it works perfectly:

```diff
diff --git a/ajax/dashboard.php b/ajax/dashboard.php
index 9455432e10..5ed708bb9d 100644
--- a/ajax/dashboard.php
+++ b/ajax/dashboard.php
@@ -146,6 +146,7 @@ switch ($_GET['action'] ?? null) {
             exit();
         }

+        header("Content-Type: application/json; charset=UTF-8");
         echo $dashboard->getFilter();
         exit;
 }
diff --git a/js/modules/Dashboard/Dashboard.js b/js/modules/Dashboard/Dashboard.js
index 6e611fc35d..fd0a3f3f9a 100644
--- a/js/modules/Dashboard/Dashboard.js
+++ b/js/modules/Dashboard/Dashboard.js
@@ -1282,7 +1282,7 @@ class GLPIDashboard {
                 dashboard: this.current_name,
             },
             success: function(response) {
-                filters = JSON.parse(response);
+                filters = response;
             }
         });

```

I guess I could open a PR for them, not why they are needed tho.

[the remark this answers is not in the document]
Comment (cedric-anne) [reply tests/cypress/support/commands.js:94]: Indeed, you can open a PR for this, and check if there are other occurences of `JSON.parse` in dashboards.

**Comment by AdrienClairembault** [comment]
> > One of the tests doesn't seem to pass on my side:
> 
> As long as it is using the test database and bootstrap file to load some initial data, it shouldn't be showing no tickets here. It isn't required to bootstrap initial data though; I think creating the items manually would be better for test coverage.
> 
> I don't think it is an issue to remove the tests currently in this PR anyways since they just show a proof of concept and show how testing certain elements that we had difficulty testing with Panther can be tested with Cypress.

I find it weird too, I would understand if the test failed on my machine but this is through docker script so it should not happen.

I'll try to dig into it, would be nice to know what going on here to avoid potential issues in the future.

**Comment by cedric-anne** [comment]
> > > One of the tests doesn't seem to pass on my side:
> > 
> > 
> > As long as it is using the test database and bootstrap file to load some initial data, it shouldn't be showing no tickets here. It isn't required to bootstrap initial data though; I think creating the items manually would be better for test coverage.
> > I don't think it is an issue to remove the tests currently in this PR anyways since they just show a proof of concept and show how testing certain elements that we had difficulty testing with Panther can be tested with Cypress.
> 
> I find it weird too, I would understand if the test failed on my machine but this is through docker script so it should not happen.
> 
> I'll try to dig into it, would be nice to know what going on here to avoid potential issues in the future.

If you run e2e tests only, the `tests/bootstrap.php` file will not be executed. Indeed it is only executed when lauching tests with `atoum`. Either we have to move these bootstrap data somewhere else, either we have to ensure that database is "clean" before executing e2e tests.

**Comment by AdrienClairembault** [comment]
Maybe we could add a "fake" atoum dedicated suite that would always be ran ?
The suite would contain no real test and just force the bootstrap to be loaded.

**Comment by AdrienClairembault** [comment]
> The web test suite may alter the bootstraped data, we have to take this into consideration too.

I guess in that case it should always be ran last.

**Comment by cconard96** [comment]
There is a hook for Cypress that lets you run code before any spec files are run.
https://docs.cypress.io/api/plugins/before-run-api

It might be possible to make a request to the bootstrap file to have it loaded on the server, but it would probably be blocked by the proxy router (if the apache container is configured with the public folder as the document root).

Another option includes using the API to create items as needed without the overhead of the UI.

Continuing with the API route, we can have a controller which automatically registers whenever it is in the testing environment which has endpoints for some direct server communication like loading fixtures, rolling back the DB, etc.

**Comment by cedric-anne** [comment]
Or we could also consider that our e2e test suite should be resilient and we could make tests that will always pass, even when some data has already been added in DB.

For instance, we can create a dedicated entity on test suite initialization and always use this entity to be sure to not see items from the other entitites.

**Comment by cconard96** [comment]
> Or we could also consider that our e2e test suite should be resilient and we could make tests that will always pass, even when some data has already been added in DB.
> 
> For instance, we can create a dedicated entity on test suite initialization and always use this entity to be sure to not see items from the other entitites.

That is probably the best idea.

**Comment by cconard96** [comment]
So close to finishing this, yet dashboard JS errors are the only thing still blocking me.
If anyone has suggestions, I'd be happy to take them. Clearly intercepting the AJAX requests and returning fake but valid data isn't working (maybe I'm doing it wrong).
When trying to visit the URL required to change profiles, the dashboard is still loading from the current page and seems to cause invalid data to be received (not sure where the HTML is coming from). By default, Cypress will fail tests when JS errors occur.

On a related note, I added a CI step to run on failure to upload any screenshots from Cypress as artifacts. You can see them by going to the details, clicking Summary on the left side and then scrolling to the bottom of the page. Its also found on the Checks tab at the top, labeled "Artifacts".

**Comment by cconard96** [comment]
🎊 Finally all tests pass.
 
 E2E tests now work regardless of which order the test suites run in. If you run e2e locally and then try running unit or functional tests afterwards, the database will be re-initialized since those other tests aren't written to take into account extra data.
 E2E tests are also partially sandboxed into its own sub-entity and uses its own user account.

**Review by AdrienClairembault — APPROVED** [review]

**Comment by trasher** [comment]
I've tried to run tests locally. I've installed a fresh DB test instance (`--config-dir=./tests/config`), launched from PHP internal server with `php -S localhost:8088 tests/router.php`, ad adapted `tests/cypress.config.js`.
Tests begin to run, but it seems it fails on login.

**Comment by cconard96** [comment]
> I've tried to run tests locally. I've installed a fresh DB test instance (`--config-dir=./tests/config`), launched from PHP internal server with `php -S localhost:8088 tests/router.php`, ad adapted `tests/cypress.config.js`. Tests begin to run, but it seems it fails on login.

The actual login fails, or it has some other issue?
The user that the tests login as is only added when GLPI is installed with the GLPI_ENVIRONMENT_TYPE set to "testing". If you installed a fresh GLPI DB without setting that as an env variable, you have a "production" database and it won't work. Installing on in the testing environment adds a new e2e_tests user (password: glpi), E2ETestEntity entity and assigns all default profiles to the e2e_tests user. Running tests from `tests/run_tests.sh` (containerized tests) automatically handles that.

**Comment by cconard96** [comment]
> Maybe we could have some `.env` config file to be able to run the tests without docker if needed ? This would allow to configure `CYPRESS_BASE_URL`, `CYPRESS_LOGIN` and `CYPRESS_PASSWORD` for each repository.
> 
> edit: some complex tests scenarios might require multiple users, not sure how we would handle that

I'm not sure it is needed to add more complexity. If someone doesn't want to use the per-configured testing environment, they can manage it themselves easily enough. There are only 2 environment variables they need to worry about.

(These commands were not tested)

Initialize e2e DB:
```sh
GLPI_ENVIRONMENT_TYPE="testing" bin/console database:install --db-host=localhost --db-name=glpi_e2e --db-user=mydbuser --db-password=mydbpassword
```

Start built-in server (router already sets env to testing):
```sh
php -S localhost:8088 tests/router.php
```

Run e2e tests:
```sh
CYPRESS_BASE_URL="localhost:8088" npm run test:e2e
```

Cypress will already take CYPRESS_ env variables to override what is in the config and GLPI also supports taking environment variables.

**Comment by trasher** [comment]
> (These commands were not tested)
> [...]
> ```shell
> CYPRESS_BASE_URL="localhost:8088" npm run test:e2e
> ```

That's better, thnak you.
This last command require schema. It was previously trying to login with glpi user, now it does with e2e_tests.

Anyway, it still got some failures :/

Under chrome, I have the following error:
```
(uncaught exception)TypeError: GLPI.Search.ResultsView is not a constructor
```

And after that, it *seems* to loop.

image

With firefox, I got:
```
(uncaught exception)ReferenceError: FullCalendarLocales is not defined
```

image

Not related, but maybe could you rebase with current main in order to run test suite with latest sources.

**Comment by trasher** [comment]
I've just tested again, and I got different errors; but unfortunately that still does not work on my local instance with both chromium and firefox :/
I'll try again next week, maybe is there something wrong on my side.

**Comment by cconard96** [comment]
It isn't just you; I see the same issues. It clearly isn't happy with the built-in PHP server.

**Comment by trasher** [comment]
> It isn't just you; I see the same issues. It clearly isn't happy with the built-in PHP server.

Hum... OK. I can setup a virtual host for tests then; that's fine for me. I'll check that next week, no need to lost time on php built-in server ;)

**Comment by cconard96** [comment]
> > It isn't just you; I see the same issues. It clearly isn't happy with the built-in PHP server.
> 
> Hum... OK. I can setup a virtual host for tests then; that's fine for me. I'll check that next week, no need to lost time on php built-in server ;)

I don't understand it, but removing the `return false` in the test router and then restarting the PHP server allows it to actually serve all of the assets that were failing with an empty response previously. Someone with more experience with the built-in server could probably create a more correct fix for the router to account for serving static assets which it didn't have to do before.

**Comment by trasher** [comment]
Modifying router file does not change anything on my side :/ I'll try with a dedicated virtualhost.

**Comment by trasher** [comment]
All tests are running fine on chrome, but all fail for firefox (with the same error as far as I've seen).

**Comment by trasher** [comment]
> All tests are running fine on chrome, but all fail for firefox (with the same error as far as I've seen).

I forgot to mention: that's not a blocker ;)

**Review by trasher — APPROVED** [review]

File: tests/cypress/e2e/poc.cy.js
[the remark this answers is not in the document]
Comment (cconard96) [reply tests/cypress/e2e/poc.cy.js:1]: Dropped POC tests

**Review by cedric-anne — APPROVED** [review]

**Pull request merged**